File: Cpp_SDL2_OpenGL_deprecated/snake.cpp

/*
TODO: SDL_ttf  length label, FPS counter

*/

#include <iostream>
#include <set>
#include <vector>
#include <unordered_map>
#include <string>
#include <random>
#include <algorithm>
#include <chrono>
#include <SDL.h>
#include <SDL_image.h>
#include <SDL_mixer.h>
#include <GL/gl.h>

#include "audiosystem.h"

constexpr int WINDOW_WIDTH = 1200, WINDOW_HEIGHT = 800;
constexpr int SPRITE_IMG_WIDTH = 16, SPRITE_IMG_HEIGHT = 16;
constexpr int SPRITE_SCALE = 5;
constexpr int SPRITE_WIDTH = SPRITE_SCALE * SPRITE_IMG_WIDTH, SPRITE_HEIGHT = SPRITE_SCALE * SPRITE_IMG_HEIGHT;
constexpr int ROWS = WINDOW_HEIGHT / SPRITE_HEIGHT, COLUMNS = WINDOW_WIDTH / SPRITE_HEIGHT;

const std::string DIR_RIGHT{"right"};
const std::string DIR_LEFT{"left"};
const std::string DIR_UP{"up"};
const std::string DIR_DOWN{"down"};

const std::vector<std::vector<std::string>> IMG_NAMES{
    {"head_up", "head_right", "head_down", "head_left"}, 
    {"tail_up", "tail_right", "tail_down", "tail_left"},
    {"turn_1", "turn_2", "turn_3", "turn_4"},
    {"vertical", "horizontal", "rabbit", "grass"}
};    

void setup_opengl( int width, int height );

auto g_img_rc = [](){
    std::unordered_map<std::string, std::pair<int, int>> img_rc;
    int i = 0;
    for(auto const& names_row : IMG_NAMES)
    {
        int j = 0;
        for(auto const& name : names_row) {
            img_rc[name] = std::pair<int, int>(i, j);
            j++;
        }
        i++;
    }
    return img_rc;
}();

struct Sprite
{
    float x, y;
    float u1, v1, u2, v2;
    Sprite(int row, int column, const std::string& img_name)
    {
        move(row, column);
        set_image(img_name);
    }
    void set_image(const std::string& img_name)
    {
        auto [i, j] = g_img_rc.at(img_name);
        u1 = 0.25f * j;
        v1 = 0.25f * (i+1);
        u2 = 0.25f * (j+1);
        v2 = 0.25f * i;
    }
    void move(int row, int column)
    {
        x = column * SPRITE_WIDTH;
        y = row * SPRITE_HEIGHT;
    }
    void draw()
    {
        //std::cout << "Draw sprite:" << x << ", " <<  y << ", " << u1 << ", " << v1 << ", " << u2 << ", " << v2 << "\n";
        glColor3f(1, 1, 1);

        glBegin(GL_QUADS);
        glTexCoord2f(u1, v1);
        glVertex2f(x, y);

        glTexCoord2f(u2, v1);
        glVertex2f(x+SPRITE_WIDTH, y);

        glTexCoord2f(u2, v2);
        glVertex2f(x+SPRITE_WIDTH, y+SPRITE_HEIGHT);

        glTexCoord2f(u1, v2);
        glVertex2f(x, y+SPRITE_HEIGHT);
        glEnd();
    }
};

struct SnakePart
{
    int row, col;
    std::string dir;
    Sprite sprite;
    SnakePart(int row, int col, const std::string& dir, const std::string& img_name):
        row(row), col(col), dir(dir), sprite(row, col, img_name)
    {}    
};

struct Rabbit
{
    int row, col;
    Sprite sprite;
    Rabbit(int row, int col):
        row(row), col(col), sprite(row, col, "rabbit")
    {}
    void move(int row_, int col_)
    {
        row = row_;
        col = col_;
        sprite.move(row, col);
    }
};
class SnakeGame
{
    std::unordered_map<std::string, GLuint> textures;
    std::vector<SnakePart> snake;
    std::vector<Sprite> background;

    float snake_move_t, snake_move_t_rem;
    std::set<std::pair<int, int>> set_all_squares;
    std::string snake_dir_next;
    Rabbit rabbit;

    std::default_random_engine rnd_generator;
public:
    SnakeGame():
        rabbit(0, 0),
        rnd_generator(std::chrono::high_resolution_clock::now().time_since_epoch().count())
    {

        for(int i=0; i<ROWS; i++) {
            for(int j=0; j<COLUMNS; j++) {
                background.push_back(Sprite(i, j, "grass"));
                set_all_squares.insert(std::pair<int, int>(i, j));
            }
        }

        reset_snake();
        place_rabbit();
    }

    void place_rabbit()
    {
        std::set<std::pair<int, int>> set_snake_squares;
        for(auto& part : snake) {
            set_snake_squares.insert(std::pair<int, int>(part.row, part.col));
        }
        std::vector<std::pair<int, int>> free_squares(ROWS*COLUMNS);

        // std::set is sorted, don't need to make a copy in std::vector and sort

        auto it=std::set_difference 
            (set_all_squares.begin(), set_all_squares.end(), 
             set_snake_squares.begin(), set_snake_squares.end(), 
             free_squares.begin());
        
        free_squares.resize(it-free_squares.begin());

        //std::cout << "place_rabbit  free_squares.size(): " << free_squares.size() << "\n";

        if(free_squares.size() > 0) {
            int i = std::uniform_int_distribution<int>(0, free_squares.size()-1)(rnd_generator);
            auto [row, col] = free_squares[i];
            //std::cout << "place_rabbit  i: " << i << "  row: " << row << "  col: " << col << "\n";
            rabbit.move(row, col);
        } else {
            reset_snake();
            place_rabbit();
        }
    }

    void reset_snake()
    {
        snake.clear();
        int i=0;
        int row = ROWS / 3, col = COLUMNS / 3;
        for(auto const& name : {"head_right", "horizontal", "tail_right"}) {
            snake.push_back(SnakePart(row, col-i, "right", name));
            i++;
        }
        snake_move_t = 0.2;
        snake_dir_next = DIR_RIGHT;
    }

    void draw()
    {
        glClear(GL_COLOR_BUFFER_BIT);
        glEnable(GL_BLEND);
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

        for(auto &sprite : background) {
            sprite.draw();
        }

        rabbit.sprite.draw();

        for(auto &part : snake) {
            part.sprite.draw();
        }
    }
    void update(float dt, AudioSystem &audio_system)
    {
        snake_move_t_rem -= dt;
        if(snake_move_t_rem <= 0) {
            snake_move_t_rem += snake_move_t;
            int row = snake[0].row;
            int col = snake[0].col;
            std::string head_dir(snake_dir_next);
            if(head_dir == DIR_LEFT) {
                col--;
            } else if(head_dir == DIR_RIGHT) {
                col++;
            } else if(head_dir == DIR_UP) {
                row++;
            }
             else if(head_dir == DIR_DOWN) {
                row--;
            }
            row %= ROWS;
            col %= COLUMNS;
            if(row < 0) {
                row = ROWS - 1;
            }
            if(col < 0) {
                col = COLUMNS - 1;
            }
            snake.insert(snake.begin(), SnakePart(row, col, head_dir, "head_"+head_dir));
            auto old_dir = snake[1].dir;
            if(head_dir == old_dir) {
                if(head_dir == DIR_LEFT || head_dir == DIR_RIGHT) {
                    snake[1].sprite.set_image("horizontal");
                } else {
                    snake[1].sprite.set_image("vertical");
                }
            } else if(old_dir == DIR_DOWN) {
                if(head_dir == DIR_LEFT) {
                    snake[1].sprite.set_image("turn_4");
                } else {
                    snake[1].sprite.set_image("turn_1");
                }
            } else if(old_dir == DIR_UP) {
                if(head_dir == DIR_LEFT) {
                    snake[1].sprite.set_image("turn_3");
                } else {
                    snake[1].sprite.set_image("turn_2");
                }
            } else if(old_dir == DIR_LEFT) {
                if(head_dir == DIR_UP) {
                    snake[1].sprite.set_image("turn_1");
                } else {
                    snake[1].sprite.set_image("turn_2");
                }
            } else if(old_dir == DIR_RIGHT) {
                if(head_dir == DIR_UP) {
                    snake[1].sprite.set_image("turn_4");
                } else {
                    snake[1].sprite.set_image("turn_3");
                }
            }

            bool rabbit_eaten = rabbit.row == row && rabbit.col == col;

            if(!rabbit_eaten) {
                snake.pop_back();
                auto &dir = (snake.end()-2)->dir;
                snake.back().sprite.set_image("tail_"+dir);
            } else {
                audio_system.play_eat();
                place_rabbit();
            }

            // test self collision
            for(auto it = snake.begin()+1; it!=snake.end(); it++) {
                if(it->col == col && it->row == row) {
                    audio_system.play_die();
                    reset_snake();
                    place_rabbit();
                    break;
                }
            }
        }
    }

    void key_input(const std::string& dir)
    {
        // enum would be better
        if(dir == DIR_LEFT) {
            if(snake[0].dir != DIR_RIGHT) {
                snake_dir_next = dir;
            }
        } else if(dir == DIR_RIGHT)
        {
            if(snake[0].dir != DIR_LEFT) {
                snake_dir_next = dir;
            }
        } else if(dir == DIR_UP)
        {
            if(snake[0].dir != DIR_DOWN) {
                snake_dir_next = dir;
            }
        }
        else if(dir == DIR_DOWN)
        {
            if(snake[0].dir != DIR_UP) {
                snake_dir_next = dir;
            }
        }
    }
};

int main(/*int argc, char *argv[]*/)
{
    if(SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_EVENTS)) {
         std::cerr << "SDL initialization failed: " << SDL_GetError() << std::endl;
         return 1;
    }

    SDL_version compiled;
    SDL_version linked;

    SDL_VERSION(&compiled);
    SDL_GetVersion(&linked);
    std::cout << "SDL Version:\n";
    std::cout << "      compiled: " << int(compiled.major) << "." << int(compiled.minor) << "." << int(compiled.patch) << "\n";
    std::cout << "        linked: " << int(linked.major) << "." << int(linked.minor) << "." << int(linked.patch) << std::endl;

    IMG_Init(IMG_INIT_PNG);

    // request opengl context attributes
    SDL_GL_SetAttribute( SDL_GL_DOUBLEBUFFER, 1 );

    SDL_Window *window = SDL_CreateWindow(
        "Snake game (C++ with SDL2/OpenGL [deprecated])", 
        SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, 
        WINDOW_WIDTH, WINDOW_HEIGHT, 
        SDL_WINDOW_ALLOW_HIGHDPI | SDL_WINDOW_OPENGL);

    if(!window) {
        std::cerr << "Cannot create window: " << SDL_GetError() << std::endl;
        SDL_Quit();
        return 1;
    }

    SDL_GLContext glcontext = SDL_GL_CreateContext(window);

    SDL_Renderer *renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);

    setup_opengl(WINDOW_WIDTH, WINDOW_HEIGHT);

    SDL_Texture *texture = IMG_LoadTexture(renderer, "../common_data/Snake.png");
    if (!texture) {
        std::cerr << "Cannot load texture:" << IMG_GetError() << std::endl;
        return 1;
    }

    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);

    SDL_GL_BindTexture(texture, NULL, NULL);

    if( Mix_OpenAudio( 44100, MIX_DEFAULT_FORMAT, 2, 2048 ) < 0 )
    {
        std::cerr << "Cannot open audio:" << Mix_GetError() << std::endl;
        return 1;
    }

    // TODO: newer SDL2 version has SDL_GetTicks64

    { // game and audio system scope
        AudioSystem audio_system;
        SnakeGame snake_game;
        Uint32 last_ticks = SDL_GetTicks();
        bool done = false;
        while(!done) {
            SDL_Event event;
            while(SDL_PollEvent(&event)) {

                switch( event.type ) {
                    case SDL_KEYDOWN:
                        std::cout << SDL_GetKeyName(event.key.keysym.sym) << std::endl;
                        switch(event.key.keysym.sym)
                        {
                            case SDLK_LEFT:  snake_game.key_input(DIR_LEFT);  break;
                            case SDLK_RIGHT: snake_game.key_input(DIR_RIGHT);  break;
                            case SDLK_UP:    snake_game.key_input(DIR_UP);  break;
                            case SDLK_DOWN:  snake_game.key_input(DIR_DOWN);  break;
                            case SDLK_ESCAPE: 
                                done = true;
                            default:
                                ;
                        }
                        break;
                    case SDL_QUIT:
                        done = true;
                    default:
                        ;
                }
            }

            Uint32 now_ticks = SDL_GetTicks();
            float dt = SDL_TICKS_PASSED(last_ticks, now_ticks) / 1000.0f;
            last_ticks = now_ticks;

            snake_game.update(dt, audio_system);

            snake_game.draw();
            SDL_GL_SwapWindow(window);
        }
    }

    SDL_GL_DeleteContext(glcontext);
    SDL_DestroyWindow(window);

    Mix_Quit();
    IMG_Quit();
    SDL_Quit();
    return 0;
}


void setup_opengl( int width, int height )
{
    glShadeModel( GL_FLAT);

    glCullFace( GL_BACK );
    glFrontFace( GL_CCW );
    glEnable( GL_CULL_FACE );

    glClearColor( 0, 0, 0, 0 );

    glViewport( 0, 0, width, height );

    glMatrixMode( GL_PROJECTION );
    glLoadIdentity( );

    glOrtho(0, width, 0, height, 1, -1);

    glMatrixMode( GL_MODELVIEW );

}

File: Java_LWJGL/Snake.java
import org.lwjgl.*;
import org.lwjgl.glfw.*;
import org.lwjgl.opengl.*;
import org.lwjgl.system.*;
import org.lwjgl.openal.*;

import java.nio.*;
import java.util.*;

import static org.lwjgl.glfw.Callbacks.*;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.system.MemoryStack.*;
import static org.lwjgl.system.MemoryUtil.*;
import static org.lwjgl.stb.STBImage.stbi_load;
import static org.lwjgl.openal.AL10.*;
import static org.lwjgl.openal.ALC11.*;


public class Snake {
    public static final int WINDOW_WIDTH = 1200, WINDOW_HEIGHT = 800;
    public static final int SPRITE_IMG_WIDTH = 16, SPRITE_IMG_HEIGHT = 16;
    public static final int SPRITE_SCALE = 5;
    public static final int SPRITE_WIDTH = SPRITE_SCALE * SPRITE_IMG_WIDTH;
    public static final int SPRITE_HEIGHT = SPRITE_SCALE * SPRITE_IMG_HEIGHT;
    public static final int ROWS = WINDOW_HEIGHT / SPRITE_HEIGHT;
    public static final int COLUMNS = WINDOW_WIDTH / SPRITE_HEIGHT;

    public static final String[][] IMG_NAMES = {
        {"head_up", "head_right", "head_down", "head_left"}, 
        {"tail_up", "tail_right", "tail_down", "tail_left"},
        {"turn_1", "turn_2", "turn_3", "turn_4"},
        {"vertical", "horizontal", "rabbit", "grass"}
    };

    public static final Map<String, int[]> IMG_RC;
    static {
        
        IMG_RC = new HashMap<>();
        int i = 0;
        for(String[] names_row : IMG_NAMES) {
            int j = 0;
            for(String name : names_row) {
                int[] rc = {i, j};
                IMG_RC.put(name, rc);
                j++;
            }
            i++;
        }
    }

    public static final String DIR_RIGHT = "right", DIR_LEFT = "left", DIR_UP = "up",  DIR_DOWN = "down";

    static class Sprite {  // nested class to avoid multiple files
        public float x, y;
        public float u1, v1, u2, v2;
        Sprite(int row, int column, String img_name)
        {
            move(row, column);
            set_image(img_name);
        }
        void set_image(String img_name)
        {
            int[] row_col = IMG_RC.get(img_name);
            int i = row_col[0], j = row_col[1];
            u1 = 0.25f * j;
            v1 = 0.25f * (i+1);
            u2 = 0.25f * (j+1);
            v2 = 0.25f * i;
        }
        void move(int row, int column)
        {
            x = column * SPRITE_WIDTH;
            y = row * SPRITE_HEIGHT;
        }
        void draw()
        {
            //std::cout << "Draw sprite:" << x << ", " <<  y << ", " << u1 << ", " << v1 << ", " << u2 << ", " << v2 << "\n";
            glColor3f(1, 1, 1);
    
            glBegin(GL_QUADS);
            glTexCoord2f(u1, v1);
            glVertex2f(x, y);
    
            glTexCoord2f(u2, v1);
            glVertex2f(x+SPRITE_WIDTH, y);
    
            glTexCoord2f(u2, v2);
            glVertex2f(x+SPRITE_WIDTH, y+SPRITE_HEIGHT);
    
            glTexCoord2f(u1, v2);
            glVertex2f(x, y+SPRITE_HEIGHT);
            glEnd();
        }
    }

    static class SnakePart {  // nested class to avoid multiple files
        public int row, col;
        public String dir;
        public Sprite sprite;
        SnakePart(int row, int col, String dir, String img_name) 
        {
            this.row = row;
            this.col = col;
            this.dir = dir;
            sprite = new Sprite(row, col, img_name);
        }
    }

    static class Rabbit {  // nested class to avoid multiple files
        public int row, col;
        public Sprite sprite;
        Rabbit(int row, int col) 
        {
            this.row = row;
            this.col = col;
            sprite = new Sprite(row, col, "rabbit");
        }
        void move(int row, int col)
        {
            this.row = row;
            this.col = col;
            sprite.move(row, col);
        }
    }

    private long window;  // window handle
    // graphics
    private int textureId;
    private Sprite[] background;
    // game logic + sprites
    private ArrayList<SnakePart> snake;
    private Rabbit rabbit;
    private String snake_dir_next;
    private float snake_move_t, snake_move_t_rem;

    private Random rnd_generator;

    private static final HashSet<Map.Entry<Integer, Integer>> set_all_squares;
    static {
        set_all_squares = new HashSet<Map.Entry<Integer, Integer>>();

        for(int i=0; i<ROWS; i++) {
            for(int j=0; j<COLUMNS; j++) {
                set_all_squares.add(Map.entry(i, j));
            }
        }
    }
    

    public void run() {
        System.out.println("LWJGL version:" + Version.getVersion());

        init();
        loop();

        // Free the window callbacks and destroy the window
        glfwFreeCallbacks(window);
        glfwDestroyWindow(window);

        // Terminate GLFW and free the error callback
        glfwTerminate();
        glfwSetErrorCallback(null).free();
    }

    private void key_pressed(long target_window, int key, int scancode, int action, int mods)
    {  // target_window should be this.window
        if(action == GLFW_PRESS) {
            String key_name = glfwGetKeyName(key, scancode);
            if(key_name == null) {
                key_name = "(a non printable key with code " + key + ")";
            }
            System.out.println("Key pressed: " + key_name);
            switch(key) {
                case GLFW_KEY_LEFT:
                    if(snake.get(0).dir != DIR_RIGHT) {
                        snake_dir_next = DIR_LEFT;
                    }
                break;
                case GLFW_KEY_RIGHT:
                    if(snake.get(0).dir != DIR_LEFT) {
                        snake_dir_next = DIR_RIGHT;
                    }
                break;
                case GLFW_KEY_UP:
                    if(snake.get(0).dir != DIR_DOWN) {
                        snake_dir_next = DIR_UP;
                    }
                break;
                case GLFW_KEY_DOWN:
                    if(snake.get(0).dir != DIR_UP) {
                        snake_dir_next = DIR_DOWN;
                    }
                break;
                case GLFW_KEY_ESCAPE:
                    glfwSetWindowShouldClose(window, true);
                break;
            }
        }
    }

    private void init() {
        System.out.println("IMG_RC:");
        for(Map.Entry<String, int[]> rc : IMG_RC.entrySet()) {
            System.out.println(
                "   " + rc.getKey() + 
                ": " + rc.getValue()[0] + ", " + rc.getValue()[1]);
        }

        rnd_generator = new Random();

        // Setup an error callback. The default implementation
        // will print the error message in System.err.
        GLFWErrorCallback.createPrint(System.err).set();

        // Initialize GLFW. Most GLFW functions will not work before doing this.
        if ( !glfwInit() )
            throw new IllegalStateException("Unable to initialize GLFW");

        // Configure GLFW
        glfwDefaultWindowHints(); // optional, the current window hints are already the default
        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE); // the window will stay hidden after creation
        glfwWindowHint(GLFW_RESIZABLE, GLFW_FALSE); // the window will be resizable

        // Create the window
        window = glfwCreateWindow(WINDOW_WIDTH, WINDOW_HEIGHT, "Snake game (Java with LWJGL)", NULL, NULL);
        if ( window == NULL )
            throw new RuntimeException("Failed to create the GLFW window");

        // Setup a key callback. It will be called every time a key is pressed, repeated or released.
		glfwSetKeyCallback(window, (window, key, scancode, action, mods) -> { key_pressed(window, key, scancode, action, mods); });

        // Get the thread stack and push a new frame
        try ( MemoryStack stack = stackPush() ) {
            IntBuffer pWidth = stack.mallocInt(1); // int*
            IntBuffer pHeight = stack.mallocInt(1); // int*

            // Get the window size passed to glfwCreateWindow
            glfwGetWindowSize(window, pWidth, pHeight);

            // Get the resolution of the primary monitor
            GLFWVidMode vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor());

            // Center the window
            glfwSetWindowPos(
                window,
                (vidmode.width() - pWidth.get(0)) / 2,
                (vidmode.height() - pHeight.get(0)) / 2
            );
        } // the stack frame is popped automatically

        // Make the OpenGL context current
        glfwMakeContextCurrent(window);
        // Enable v-sync
        glfwSwapInterval(1);

        // Make the window visible
        glfwShowWindow(window);

        GL.createCapabilities();

        load_texture();

        glShadeModel(GL_SMOOTH);
        glCullFace(GL_BACK);
        glFrontFace(GL_CCW);
        glEnable(GL_CULL_FACE);
        glMatrixMode(GL_PROJECTION);
        glLoadIdentity();
        glOrtho(0, WINDOW_WIDTH, 0, WINDOW_HEIGHT, 1, -1);
        glMatrixMode(GL_MODELVIEW);

        ArrayList<Sprite> sprites = new ArrayList<>();
        for(int i=0; i<ROWS; i++) {
            for(int j=0; j<COLUMNS; j++) {
                sprites.add(new Sprite(i, j, "grass"));
            }
        }
        background = new Sprite[sprites.size()];
        sprites.toArray(background);

        snake = new ArrayList<SnakePart>();
        reset_snake();

        rabbit = new Rabbit(0, 0);
        place_rabbit();
    }

    void reset_snake()
    {
        snake.clear();
        List<String> names = List.of("head_right", "horizontal", "tail_right");
        ListIterator<String> it = names.listIterator();
        int row = ROWS / 3, col = COLUMNS / 3;
        while(it.hasNext()) {
            snake.add(new SnakePart(row, col-it.nextIndex(), DIR_RIGHT, it.next()));
        }
        snake_dir_next = DIR_RIGHT;
        snake_move_t = 0.2f;
    }

    void place_rabbit()
    {
        HashSet<Map.Entry<Integer, Integer>> free_squares = (HashSet)set_all_squares.clone();
        snake.forEach((sp) -> free_squares.remove(Map.entry(sp.row, sp.col)));
        
        if(free_squares.size() == 0) {
            System.out.println("victory !!!");
            reset_snake();
            place_rabbit();
        } else {
            ArrayList<int[]> free_squares_array = new ArrayList<int[]>();
            free_squares.forEach((fs) -> free_squares_array.add(new int[]{fs.getKey(), fs.getValue()}));
            int i = rnd_generator.nextInt(free_squares.size());
            int[] rc = free_squares_array.get(i);
            rabbit.move(rc[0], rc[1]);
        }
    }

    void update(float dt)
    {
        snake_move_t_rem -= dt;
        if(snake_move_t_rem <= 0) {
            snake_move_t_rem += snake_move_t;
            int row = snake.get(0).row;
            int col = snake.get(0).col;
            String head_dir = snake_dir_next;
            if(head_dir == DIR_LEFT) {
                col--;
            } else if(head_dir == DIR_RIGHT) {
                col++;
            } else if(head_dir == DIR_UP) {
                row++;
            } else if(head_dir == DIR_DOWN) {
                row--;
            }
            row %= ROWS;
            col %= COLUMNS;
            if(row < 0) {
                row = ROWS - 1;
            }
            if(col < 0) {
                col = COLUMNS - 1;
            }
            
            { // insert new head to front
                SnakePart sp = new SnakePart(row, col, head_dir, "head_"+head_dir);
                snake.add(0, sp);
            }

            String old_dir = snake.get(1).dir;
            if(head_dir == old_dir) {
                if(head_dir == DIR_LEFT || head_dir == DIR_RIGHT) {
                    snake.get(1).sprite.set_image("horizontal");
                } else {
                    snake.get(1).sprite.set_image("vertical");
                }
            } else if(old_dir == DIR_DOWN) {
                if(head_dir == DIR_LEFT) {
                    snake.get(1).sprite.set_image("turn_4");
                } else {
                    snake.get(1).sprite.set_image("turn_1");
                }
            } else if(old_dir == DIR_UP) {
                if(head_dir == DIR_LEFT) {
                    snake.get(1).sprite.set_image("turn_3");
                } else {
                    snake.get(1).sprite.set_image("turn_2");
                }
            } else if(old_dir == DIR_LEFT) {
                if(head_dir == DIR_UP) {
                    snake.get(1).sprite.set_image("turn_1");
                } else {
                    snake.get(1).sprite.set_image("turn_2");
                }
            } else if(old_dir == DIR_RIGHT) {
                if(head_dir == DIR_UP) {
                    snake.get(1).sprite.set_image("turn_4");
                } else {
                    snake.get(1).sprite.set_image("turn_3");
                }
            }

            Boolean rabbit_eaten = rabbit.row == row && rabbit.col == col;

            if(!rabbit_eaten) {
                snake.remove(snake.size()-1);
                String dir = (snake.get(snake.size()-2)).dir;
                snake.get(snake.size()-1).sprite.set_image("tail_"+dir);
            } else {
                //audio_system.play_eat();
                place_rabbit();
            }            

            // test self collision
            Iterator<SnakePart> spIt = snake.listIterator(1);
            while(spIt.hasNext()) {
                SnakePart sp = spIt.next();
                if(sp.col == col && sp.row == row) {
                    //audio_system.play_die();
                    reset_snake();
                    place_rabbit();
                    break;
                }
            }
        }
    }
    
    private void loop() {
        glClearColor(0.0f, 0.0f, 0.0f, 0.0f);

        glfwSetTime(0.0);
        while ( !glfwWindowShouldClose(window) ) {
            float dt = (float)glfwGetTime();
            glfwSetTime(0.0); // reset the timer every frame

            update(dt);

            glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT); // clear the framebuffer

            glEnable(GL_TEXTURE_2D);

            for(Sprite sprite : background) {
                sprite.draw();
            }

            glEnable(GL_BLEND);
            glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    
            for(SnakePart part : snake) {
                part.sprite.draw();
            }
            
            rabbit.sprite.draw();

            glDisable(GL_BLEND);

            glfwSwapBuffers(window); // swap the color buffers

            // Poll for window events. The key callback above will only be
            // invoked during this call.
            glfwPollEvents();
        }
    }

    private void load_texture()
    {
        textureId = glGenTextures();
        glBindTexture(GL_TEXTURE_2D, textureId);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        IntBuffer width = BufferUtils.createIntBuffer(1);
        IntBuffer height = BufferUtils.createIntBuffer(1);
        IntBuffer channels = BufferUtils.createIntBuffer(1);
        ByteBuffer image = stbi_load("../common_data/Snake.png", width, height, channels, 0);
        if (image == null) {
            throw new RuntimeException("stbi_load failed");
        }
        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width.get(0), height.get(0), 0, GL_RGBA, GL_UNSIGNED_BYTE, image);
    }

    public static void main(String[] args) {
        long device = alcOpenDevice(args.length == 0 ? null : args[0]);
        if (device == NULL) {
            throw new IllegalStateException("Failed to open an OpenAL device.");
        }
                
        new Snake().run();
    }

}
